From 2fd7b46f1c36569b5b8752ed7ff2e1e3d89e4be1 Mon Sep 17 00:00:00 2001
From: Patrick Delaunay <patrick.delaunay@foss.st.com>
Date: Fri, 19 Jan 2024 17:53:25 +0100
Subject: [PATCH 549/954] feat(stm32mp2): access to DDR regulators by supply
 name

Remove assumption of DDR supply name, and get the DDR supply
defined in device tree.

The regulator can be provided by STPMIC25 or any other regulator.

For LPDDR4, this patch adds the support VDDQ even if on ST32MP25 boards
VDD2 = VDDQ = buck6 of STPMIC25

For DDR4, this patch removes the STPMIC2 specific configuration for
REGUL_SINK_SOURCE. It is now handled in device tree.

Upstream-Status: Inappropriate [ST downstream https://github.com/STMicroelectronics/arm-trusted-firmware.git 38d275feae929425c41dbb1c00cfc8f86c087896 feat(stm32mp2): access to DDR regulators by supply name]
Change-Id: Idfdd47e500b0f57c9db9779e5a15322e1021849a
Signed-off-by: Patrick Delaunay <patrick.delaunay@foss.st.com>
Reviewed-on: https://gerrit.st.com/c/mpu/oe/st/tf-a/+/355239
Reviewed-on: https://gerrit.st.com/c/mpu/oe/st/tf-a/+/368533
ACI: CITOOLS <MDG-smet-aci-reviews@list.st.com>
ACI: CIBUILD <MDG-smet-aci-builds@list.st.com>
Reviewed-by: Yann GAUTIER <yann.gautier@foss.st.com>
---
 plat/st/stm32mp2/plat_ddr.c | 168 ++++++++++++++++++++----------------
 1 file changed, 92 insertions(+), 76 deletions(-)

diff --git a/plat/st/stm32mp2/plat_ddr.c b/plat/st/stm32mp2/plat_ddr.c
index 004a5554c7..05923e04c5 100644
--- a/plat/st/stm32mp2/plat_ddr.c
+++ b/plat/st/stm32mp2/plat_ddr.c
@@ -6,127 +6,127 @@
 
 #include <assert.h>
 #include <errno.h>
+#include <libfdt.h>
 #include <stdint.h>
 
+#include <common/fdt_wrappers.h>
 #include <drivers/delay_timer.h>
+#include <drivers/st/regulator.h>
 #include <drivers/st/stm32mp_ddr.h>
-#include <drivers/st/stm32mp_pmic2.h>
 
 #include <platform_def.h>
 
-static int pmic_ddr_power_init(void)
-{
 #if STM32MP_DDR4_TYPE
+struct ddr4_supply {
+	struct rdev *vdd;
+	struct rdev *vref;
+	struct rdev *vtt;
+	struct rdev *vpp;
+};
+
+static void ddr4_supply_read(void *fdt, int node, struct ddr4_supply *supply)
+{
+	supply->vpp = regulator_get_by_supply_name(fdt, node, "vpp");
+	supply->vdd = regulator_get_by_supply_name(fdt, node, "vdd");
+	supply->vref = regulator_get_by_supply_name(fdt, node, "vref");
+	supply->vtt = regulator_get_by_supply_name(fdt, node, "vtt");
+}
+
+static int ddr_power_init(void *fdt, int node)
+{
 	int status;
-	struct rdev *vdd_ddr, *vref_ddr, *vtt_ddr, *vpp_ddr;
+	struct ddr4_supply supply;
+
+	ddr4_supply_read(fdt, node, &supply);
+	if ((supply.vpp == NULL) || (supply.vdd == NULL) || (supply.vref == NULL) ||
+	    (supply.vtt == NULL)) {
+		return -ENOENT;
+	}
 
 	/*
-	 * DDR power on sequence is:
+	 * DDR4 power on sequence is:
 	 * enable VPP_DDR
 	 * wait 2ms
 	 * enable VREF_DDR, VTT_DDR, VPP_DDR
 	 */
-
-	/*
-	 * FIXME: to be replaced by regulator_get_by_supply_name(... "vpp")
-	 * vpp-supply = <&vpp_ddr>; in the device-tree
-	 */
-	/* vpp_ddr */
-	vpp_ddr = regulator_get_by_name("ldo5");
-	if (vpp_ddr == NULL) {
-		return -ENOENT;
-	}
-	status = regulator_set_min_voltage(vpp_ddr);
-	if (status != 0) {
-		return status;
-	}
-
-	/* vdd_ddr */
-	vdd_ddr = regulator_get_by_name("buck6");
-	if (vdd_ddr == NULL) {
-		return -ENOENT;
-	}
-	status = regulator_set_min_voltage(vdd_ddr);
+	status = regulator_set_min_voltage(supply.vpp);
 	if (status != 0) {
 		return status;
 	}
 
-	/* vref_ddr */
-	vref_ddr = regulator_get_by_name("refddr");
-	if (vref_ddr == NULL) {
-		return -ENOENT;
-	}
-
-	/* vtt_ddr */
-	vtt_ddr = regulator_get_by_name("ldo3");
-	if (vtt_ddr == NULL) {
-		return -ENOENT;
-	}
-	status = regulator_set_flag(vtt_ddr, REGUL_SINK_SOURCE);
+	status = regulator_set_min_voltage(supply.vdd);
 	if (status != 0) {
 		return status;
 	}
 
-	status = regulator_enable(vpp_ddr);
+	status = regulator_enable(supply.vpp);
 	if (status != 0) {
 		return status;
 	}
 
 	udelay(1000);
 
-	status = regulator_enable(vdd_ddr);
+	status = regulator_enable(supply.vdd);
 	if (status != 0) {
 		return status;
 	}
 
-	status = regulator_enable(vref_ddr);
+	status = regulator_enable(supply.vref);
 	if (status != 0) {
 		return status;
 	}
 
-	status = regulator_enable(vtt_ddr);
-	if (status != 0) {
-		return status;
-	}
+	return regulator_enable(supply.vtt);
+}
+#endif /* STM32MP_DDR4_TYPE */
 
-	udelay(1000);
+#if STM32MP_LPDDR4_TYPE
+struct lpddr4_supply {
+	struct rdev *vdd1;
+	struct rdev *vdd2;
+	struct rdev *vddq;
+};
 
-#elif STM32MP_LPDDR4_TYPE
+static void lpddr4_supply_read(void *fdt, int node, struct lpddr4_supply *supply)
+{
+	supply->vdd1 = regulator_get_by_supply_name(fdt, node, "vdd1");
+	supply->vdd2 = regulator_get_by_supply_name(fdt, node, "vdd2");
+	supply->vddq = regulator_get_by_supply_name(fdt, node, "vddq");
+}
+
+static int ddr_power_init(void *fdt, int node)
+{
 	int status;
-	struct rdev *vdd1_ddr, *vdd2_ddr;
+	struct lpddr4_supply supply;
+
+	lpddr4_supply_read(fdt, node, &supply);
+	if ((supply.vdd1 == NULL) || (supply.vdd2 == NULL) || (supply.vddq == NULL)) {
+		return -ENOENT;
+	}
 
 	/*
-	 * DDR power on sequence is:
+	 * LPDDR4 power on sequence is:
 	 * enable VDD1_DDR
 	 * wait 2ms
 	 * enable VDD2_DDR
+	 * enable VDDQ_DDR
 	 */
-
-	/*
-	 * FIXME: to be replaced by regulator_get_by_supply_name(... "vdd1_ddr")
-	 * vdd1-supply = <&vdd1_ddr>; in the device-tree
-	 */
-	/* vdd1_ddr */
-	vdd1_ddr = regulator_get_by_name("ldo3");
-	if (vdd1_ddr == NULL) {
-		return -ENOENT;
-	}
-	status = regulator_set_min_voltage(vdd1_ddr);
+	status = regulator_set_min_voltage(supply.vdd1);
 	if (status != 0) {
 		return status;
 	}
 
-	/* vdd2_ddr */
-	vdd2_ddr = regulator_get_by_name("buck6");
-	if (vdd2_ddr == NULL) {
-		return -ENOENT;
+	status = regulator_set_min_voltage(supply.vdd2);
+	if (status != 0) {
+		return status;
 	}
-	status = regulator_set_min_voltage(vdd2_ddr);
+
+	status = regulator_set_min_voltage(supply.vddq);
 	if (status != 0) {
 		return status;
 	}
 
-	status = regulator_enable(vdd1_ddr);
+	status = regulator_enable(supply.vdd1);
 	if (status != 0) {
 		return status;
 	}
@@ -134,24 +134,40 @@ static int pmic_ddr_power_init(void)
 	/* could be set via enable_ramp_delay on vdd1_ddr */
 	udelay(2000);
 
-	status = regulator_enable(vdd2_ddr);
+	status = regulator_enable(supply.vdd2);
 	if (status != 0) {
 		return status;
 	}
-#else
-	ERROR("DDR type no supported with PMIC\n");
-	panic();
-#endif
 
-	return 0;
+	return regulator_enable(supply.vddq);
 }
+#endif /* STM32MP_LPDDR4_TYPE */
 
 int stm32mp_board_ddr_power_init(enum ddr_type ddr_type)
 {
-	if (dt_pmic_status() > 0) {
-		VERBOSE("PMIC DDR power init, ddr_type = %u\n", ddr_type);
-		return pmic_ddr_power_init();
+	void *fdt = NULL;
+	int node;
+
+	VERBOSE("DDR power init, ddr_type = %u\n", ddr_type);
+
+#if STM32MP_DDR4_TYPE
+	assert(ddr_type == STM32MP_DDR4);
+#elif STM32MP_LPDDR4_TYPE
+	assert(ddr_type == STM32MP_LPDDR4);
+#else
+	ERROR("DDR type (%u) not supported\n", ddr_type);
+	panic();
+#endif
+
+	if (fdt_get_address(&fdt) == 0) {
+		return -FDT_ERR_NOTFOUND;
+	}
+
+	node = fdt_node_offset_by_compatible(fdt, -1, DT_DDR_COMPAT);
+	if (node < 0) {
+		ERROR("%s: Cannot read DDR node in DT\n", __func__);
+		return -EINVAL;
 	}
 
-	return 0;
+	return ddr_power_init(fdt, node);
 }
-- 
2.53.0

